internal/cmd/oohelperd/http.go

```diff
@@ -102,13 +106,56 @@ func httpDo(ctx context.Context, config *httpConfig) {
 	reader := &io.LimitedReader{R: resp.Body, N: config.MaxAcceptableBody}
 	data, err := netxlite.ReadAllContext(ctx, reader)
 	ol.Stop(err)
+
+	h3Endpoint := ""
+	if !config.h3 {
+		h3Endpoint = getHTTP3Altsvc(resp)
+	}
+
 	config.Out <- ctrlHTTPResponse{
-		BodyLength: int64(len(data)),
-		Failure:    httpMapFailure(err),
-		StatusCode: int64(resp.StatusCode),
-		Headers:    headers,
-		Title:      measurexlite.WebGetTitle(string(data)),
+		BodyLength:   int64(len(data)),
+		DiscoveredH3: h3Endpoint,
+		Failure:      httpMapFailure(err),
+		StatusCode:   int64(resp.StatusCode),
+		Headers:      headers,
+		Title:        measurexlite.WebGetTitle(string(data)),
+	}
+}
+
+func getHTTP3Altsvc(resp *http.Response) string {
+	var responses []*http.Response
+	var request *http.Request
+
+	// go through chain of HTTP Request - Responses to find first Response
+	// TODO (kelmenhorst): this will not work for http:// URLs
+	for i := 0; i < 10; i++ {
```

Comment:
10 is more or less a magic number. I had in the back of my mind that we used 10 to limit the number of redirects elsewhere. Do you propose another number? Or do you think it is safe to just loop until the break-condition holds (i.e. without limit)?

Reply:
> 10 is more or less a magic number. I had in the back of my mind that we used 10 to limit the number of redirects elsewhere [...]

A-ha, indeed! It's the number [used by default by Go's stdlib](https://github.com/golang/go/blob/go1.19.3/src/net/http/client.go#L76).

> Do you propose another number? Or do you think it is safe to just loop until the break-condition holds (i.e. without limit)?

Good point! I don't see a reason for us to request the stdlib to perform more than 10 redirects. If the goal is really to find the _first_ request in the chain, then maybe a defensive algorithm could be to iterate 11 times (i.e., one time more than what the stdlib would do) and return `nil, false` if we cannot find the first request. If this is a separate function we can instead just return `resp, true` in place when we find a match.

Your questions leads us to ask the question of what we are actually doing here. We discussed this in another comment, I think, that maybe we want to discover all the possible Alt-Svc bindings implied by the data rather than just focusing on the last one. I think that, if we have a function to extract the Alt-Svc binding from a request-response pair, we will be able to use it even when we further refactor this code to use the step-by-step style.

All in all, I think it would be perfectly okay for the first implementation of this feature to find the first request in the chain and ignore other requests. We will document that this is a known limitation and focus on rolling this improvement out sometime around next week so we can start getting the data and using it.

How does that sounds?

Reply:
Sounds good. So, I'll change the loop limit, adjust the return values of `getFirstResponseInRedirectChain` and document the limitation.